src/box/replication.cc

```diff
@@ -892,8 +997,11 @@ applier_on_connect_f(struct trigger *trigger, void *event)
 
 void
 replicaset_connect(struct applier **appliers, int count,
-		   bool connect_quorum, bool keep_connect)
+		   bool connect_quorum, bool keep_connect,
+		   struct replica **bootstrap_leader)
 {
+	if (bootstrap_leader != NULL)
+		*bootstrap_leader = NULL;
```

Comment:
It's only there for the case when `count == 0` and we might return an uninitialized `bootstrap_leader`. So I moved the check there.

<details><summary>Diff in first commit:</summary>

```diff
diff --git a/src/box/replication.cc b/src/box/replication.cc
index d126b1fbe..a7d782eb0 100644
--- a/src/box/replication.cc
+++ b/src/box/replication.cc
@@ -966,11 +966,11 @@ replicaset_connect(struct applier **appliers, int count,
                   bool connect_quorum, bool keep_connect,
                   struct replica **bootstrap_leader)
 {
-       if (bootstrap_leader != NULL)
-               *bootstrap_leader = NULL;
        if (count == 0) {
                /* Cleanup the replica set. */
                replicaset_update(appliers, 0, false);
+               if (bootstrap_leader != NULL)
+                       *bootstrap_leader = NULL;
                return;
        }
 
```
</details>

I also decided to set `bootstrap_leader` to `NULL` explicitly when it can't be found in the last commit.
<details><summary>Diff in last commit:</summary>

```diff
diff --git a/src/box/replication.cc b/src/box/replication.cc
index d740c03bc..11af1d094 100644
--- a/src/box/replication.cc
+++ b/src/box/replication.cc
@@ -1119,15 +1119,26 @@ replicaset_connect(struct applier **appliers, int count,
        } catch (Exception *e) {
                goto error;
        }
-       if (bootstrap_leader != NULL) {
-               if (state.bootstrap_leader_applier != NULL) {
+
+       if (bootstrap_leader == NULL)
+               return;
+
+       switch (bootstrap_strategy) {
+       case BOOTSTRAP_STRATEGY_LEGACY:
+       case BOOTSTRAP_STRATEGY_AUTO:
+               *bootstrap_leader = replicaset_find_join_master();
+               break;
+       case BOOTSTRAP_STRATEGY_CONFIG:
+               if (state.bootstrap_leader_applier == NULL) {
+                       *bootstrap_leader = NULL;
+               } else {
                        struct applier *leader = state.bootstrap_leader_applier;
                        *bootstrap_leader = replica_by_uuid(&leader->uuid);
                        assert(bootstrap_leader != NULL);
-               } else if (bootstrap_strategy == BOOTSTRAP_STRATEGY_AUTO ||
-                          bootstrap_strategy == BOOTSTRAP_STRATEGY_LEGACY) {
-                       *bootstrap_leader = replicaset_find_join_master();
                }
+               break;
+       default:
+               unreachable();
        }
        return;
 error:

```
</details>

Reply:
Thanks for explanation! Makes sense now.